### PR TITLE
[FIRRTL] Expand AnnotationSet mutation API

### PR DESCRIPTION
**Pull request by fabianschuiki**
* Add constructors for `AnnotationSet` directly from arrays of `Attribute` and `Annotation`. This can come in handy in certain
  situations where a user might have already filtered a bunch of attributes.

* Add a bunch of methods to apply the annotations recorded in an `AnnotationSet` to the attributes of an operation or port through various intermediate forms (directly on the op, updating a `DictionaryAttr`, modifying a `NamedAttrList`).

* Add a bunch of methods to add and remove annotations in a set. This is a convenient way for performing single-shot modifications of an op's or port's annotations (but may become inefficient if a lot of incremental changes are performed).

* Add a `removeAnnotations` method that takes a predicate function that gets applied to every annotation and can indicate whether the annotation should be retained in the set or not. This is guaranteed to be called on every annotation, and thus lends itself handily to extract the annotations consumed in a transformation.

* Make `Annotation` comparable through `==` and `!=`.

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLAnnotations.cpp [hunk 1]
```diff
@@ -17,11 +17,32 @@ using namespace circt;
...
+  if (auto annots = op->getAttrOfType<ArrayAttr>(getAnnotationAttrName()))
     return annots;
   return ArrayAttr::get(op->getContext(), {});
 }
 
+static ArrayAttr getAnnotationsFrom(ArrayRef<Annotation> annotations,
+                                    MLIRContext *context) {
+  if (annotations.empty())
+    return ArrayAttr::get(context, {});
+  SmallVector<Attribute, 8> attrs;
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLAnnotations.cpp:29]: Is there a good reason to choose `8` as opposed to just `SmallVector<Attribute>`?

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h [hunk 2]
```diff
@@ -63,6 +76,40 @@ class AnnotationSet {
   DictionaryAttr
   getArgumentAttrDict(ArrayRef<NamedAttribute> otherPortAttrs = {}) const;
 
+  /// Store this annotation set on an operation. If `removeOnEmpty` is set, the
+  /// annotation attribute is removed from the operation altogether. Returns
+  /// true if the operation was modified, false otherwise.
+  bool updateOperation(Operation *op, bool removeOnEmpty = true) const;
```
[the remark this answers is not in the document]
Comment (seldridge) [reply include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:82]: The `removeOnEmpty` parameter seems like a bit of a weird option. It seems like there should be a policy of "everything has an `"annotations"` attribute" or "empty annotation attributes aren't stored". The `AnnotationSet` constructor will already do the right thing when `"annotations"` doesn't exist, so maybe it's fine to remove `removeOnEmpty` and make this the policy for how this is handled?

**Review by seldridge — APPROVED** [review]
This looks great and I'm looking forward to using it. :100:

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h [hunk 3]
```diff
@@ -23,16 +23,29 @@ namespace firrtl {
 class Annotation;
 class AnnotationSetIterator;
 
+/// Return the name of the attribute used for annotations on FIRRTL ops.
+inline StringRef getAnnotationAttrName() { return "annotations"; }
+
+/// Return the name of the dialect-prefixed attribute used for annotations.
+inline StringRef getDialectAnnotationAttrName() { return "firrtl.annotations"; }
```
[the remark this answers is not in the document]
Comment (lattner) [reply include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:30]: A string literal lives in global immutable memory, so it is always ok to pass around a StringRef to it.

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (lattner) [reply include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:82]: How about `annotSet.applyToOperation(op);`?

Out of curiosity, why do we need the `removeOnEmpty` parameter?  Isn't this a function of the operation, so this could just do the right thing depending on what the operation is?

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (seldridge) [reply include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:30]: Got it! That makes sense.

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h [hunk 4]
```diff
@@ -63,6 +76,40 @@ class AnnotationSet {
...
+  /// Store this annotation set on an operation. If `removeOnEmpty` is set, the
+  /// annotation attribute is removed from the operation altogether. Returns
+  /// true if the operation was modified, false otherwise.
+  bool updateOperation(Operation *op, bool removeOnEmpty = true) const;
+
+  /// Store this annotation set in the attribute list of an operation. If
+  /// `removeOnEmpty` is set, the annotation attribute is removed from the
+  /// operation altogether if this set is empty. Returns true if the attribute
+  /// list was modified, false otherwise.
+  bool updateAttrs(NamedAttrList &attrs, bool removeOnEmpty = true) const;
```
Comment (lattner) [inline include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:88]: This comment is out of date.  Please also describe what `removeOnEmpty` does here.

More importantly, it is also not obvious to me that this is setting the "annotations" attribute in `attrs` with the attribute set.  Could we make this be more port specific somehow?  The version immediately below looks more directed and obvious how to use it.  Do we need this one?

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h [hunk 5]
```diff
@@ -63,6 +76,40 @@ class AnnotationSet {
...
+  /// altogether if this set is empty. Returns true if the attribute list was
+  /// modified, false otherwise.
+  bool updatePortAttrs(NamedAttrList &attrs, bool removeOnEmpty = true) const;
+
+  /// Update the attribute dictionary of an operation to contain this annotation
+  /// set. If `removeOnEmpty` is set, the annotation attribute is removed from
+  /// the operation altogether if this set is empty. Returns the updated
+  /// dictionary.
+  DictionaryAttr updateDictionaryAttr(DictionaryAttr attrs,
+                                      bool removeOnEmpty = true) const;
```
Comment (lattner) [inline include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:101]: I'd love to get rid of `removeOnEmpty`.  Is there any way to make this just do the right thing?  Why is this needed?

File: lib/Dialect/FIRRTL/FIRRTLAnnotations.cpp [hunk 6]
```diff
@@ -119,6 +273,53 @@ void AnnotationSet::addAnnotations(ArrayAttr newAnnotations) {
...
+  // Build a filtered list of annotations.
+  SmallVector<Attribute, 8> filteredAnnos;
+  filteredAnnos.reserve(annos.size());
+  filteredAnnos.append(annos.begin(), it);
+  ++it;
+  while (it != annos.end()) {
+    if (!predicate(Annotation(it->cast<DictionaryAttr>())))
+      filteredAnnos.push_back(*it);
+    ++it;
+  }
```
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLAnnotations.cpp:318]: I'd personally just copy the attributes into a SmallVector, then use the 
` out.erase(std::remove_if(begin(out), end(out), std::not_fn(p)), end(out));`

idiom.  YMMV

**Review by lattner — COMMENTED** [review]
This is looking good, but I'm a bit concerned about the kitchen sink nature of this.  Is it possible to cut down the API surface area at all?  I'd really love to get rid of the "removeOnEmpty" parameters.

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h [hunk 7]
```diff
@@ -107,9 +154,27 @@ class AnnotationSet {
...
+  /// Remove an annotation from this annotation set. Returns true if any were
+  /// removed, false otherwise.
+  bool removeAnnotation(Annotation anno);
+  bool removeAnnotation(Attribute anno);
+
+  /// Remove all annotations from this annotation set for which `predicate`
+  /// returns true. The predicate is guaranteed to be called on every
+  /// annotation, such that this method can be used to partition the set by
+  /// extracting and removing annotations at the same time.
+  bool removeAnnotations(llvm::function_ref<bool(Annotation)> predicate);
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:171]: River of MLIR upstream scolded me for using `std::function` in places like these, which seems to very happily allocate memory on the heap (for packaging up lambdas), whereas `llvm::function_ref` apparently does a better job at avoiding this.

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:82]: Nice, I also prefer to get rid of the `removeOnEmpty` parameter and remove empty `annotations = {}` attributes. Otherwise we'll end up with a lot of these dead entries around. Just left the parameter there to leave this to the user, but actually enforcing this behaviour as part of the `AnnotationSet` API is even better!

File: lib/Dialect/FIRRTL/FIRRTLAnnotations.cpp (hunk 1 above)
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/FIRRTLAnnotations.cpp:29]: Nah good point, I'll just use the default here.

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h (hunk 4 above)
Comment (fabianschuiki) [reply include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:88]: The argument to have this as well was that if you are in the process of doing multiple modifications of an operation's attributes through a `NamedAttrList` (or you are building one), this would allow you to operate on that list directly, without having to first intern it as a `DictionaryAttr` just to modify it again through `updateDictionaryAttr` again immediately after.

I would love to see some attribute mutation classes in MLIR in general. Right now you can choose between `NamedAttrList`, rolling your own vector of `NamedAttribute`s and friends, with various degrees of not-modified-optimizations. Since the attributes are essentially an immutable data structure, a lot of the mutation API idioms we have for those could apply. Like a `MutableAttribute` helper to allow attributes to be mutated multiple types (and maybe in a nested fashion) before interning them into the context again :thinking:.

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h (hunk 5 above)
Comment (fabianschuiki) [reply include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h:101]: I'll drop that. Pretty sure we want the `removeOnEmpty = true` behaviour in all cases anyway (to not leave empty `annotations = []` behind).

**Comment by fabianschuiki** [comment]
Dropped the `removeOnEmpty` parameter (always removing on empty), renamed the update functions to `applyTo...`, improved the comments to more clearly communicate what the functions do (and when to use them), and incorporated feedback.

**Review by lattner — APPROVED** [review]
Thanks for simplifying this
